### ISSUE TITLE
Add insert/update/delete support to vaadin grid

### ISSUE THREAD

**Issue by Jonno63** [vaadin/vaadin-grid#1526]
### Description
Currently when using dataProvider to load grid data, there is no native support for insert/update/delete operations on grid data.

While,this can be achieved when using the items array to load data, this vaadin forum thread, https://vaadin.com/forum/thread/17445015/17445638, notes that the items array is intended for use with small numbers of records when the entire data set can be loaded directly.

The thread proposes a workaround but the grid would greatly benefit from native support for these operation in dataProvider mode thus retaining the lazy loading essential when dealing with a large number of records.

This thread also notes the refreshItem(T item) method of the Java component for updating items within a grid so perhaps something similar would work with corresponding methods to add and delete records too

In summary the feature request is to provide native support within grid methods to add,update  and delete individual records and have these changes reflected in the UI.

**Comment by web-padawan** [comment]
The custom logic can be implemented to handle that, and we actually have 3 ways for editing data displayed as a grid at this time (all of those are built using different approaches):

- [vaadin-crud](https://github.com/vaadin/vaadin-crud) is a web component providing the dialog as a container for the grid editor, with a form. See [Data Provider](https://cdn.vaadin.com/vaadin-crud/1.0.1/demo/#crud-data-demos) demos for the example of how it works.
- [vaadin-grid-pro](https://github.com/vaadin/vaadin-grid-pro) is another web component with the inline editing feature, where each cell of the specific editable column can be edited separately.
- [Vaadin Grid for Flow](https://vaadin.com/components/vaadin-grid/java-examples/grid-editor) has an editor feature available in the Flow counterpart only (not in the web component). There you can edit whole rows (so that's the difference from the Pro Grid).

The Pro web components are still not patching `dataProvider` logic internally but provide the events which the user could use to handle the application state updates accordingly.

So, we understand your feature request but in general there are ways to accomplish this using Vaadin platform now (although 2 of those are Pro components and require a subscription).

**Comment by lycandroid** [comment]
I'm not sure this answers the question.  The question relates to arbitrary updates of data in the grid which do not necessarily occur through a user action.  For example a server push may want to update a grid row. 

Even if it is a user action it may be in some other part of the UI.  Perhaps vaadin-crud can be used for that but does not really fit into a "pick-and-mix" web component environment where the "other part of the UI" may be a non-vaadin element.
